arwen/contexts/runtime.go

```diff
@@ -74,56 +74,134 @@ func (context *runtimeContext) InitState() {
 	}
 }
 
-func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uint64) error {
-	if context.RunningInstancesCount() >= context.maxWasmerInstances {
-		context.instance = nil
-		return arwen.ErrMaxInstancesReached
-	}
-
+func (context *runtimeContext) setWarmInstanceWhenNeeded() bool {
 	scAddress := context.GetSCAddress()
 	useWarm := context.useWarmInstance && context.warmInstanceAddress != nil && bytes.Equal(scAddress, context.warmInstanceAddress)
 	if scAddress != nil && useWarm {
 		log.Trace("Reusing the warm Wasmer instance")
 
 		context.instance = context.warmInstance
 		context.SetPointsUsed(0)
-	} else {
-		log.Trace("Creating a new Wasmer instance")
-
-		gasSchedule := context.host.Metering().GasSchedule()
-		options := wasmer.CompilationOptions{
-			GasLimit:           gasLimit,
-			UnmeteredLocals:    uint64(gasSchedule.WASMOpcodeCost.LocalsUnmetered),
-			OpcodeTrace:        false,
-			Metering:           true,
-			RuntimeBreakpoints: true,
-		}
-		newInstance, err := wasmer.NewInstanceWithOptions(contract, options)
+
+		context.SetRuntimeBreakpointValue(arwen.BreakpointNone)
+		return true
+	}
+
+	return false
+}
+
+func (context *runtimeContext) makeInstanceFromCompiledCode(codeHash []byte, gasLimit uint64, newCode bool) bool {
+	if !context.host.IsAheadOfTimeCompileEnabled() {
+		return false
+	}
+
+	if newCode || len(codeHash) == 0 {
+		return false
+	}
+
+	blockchain := context.host.Blockchain()
+	found, compiledCode := blockchain.GetCompiledCode(codeHash)
+	if !found {
+		return false
+	}
+
+	gasSchedule := context.host.Metering().GasSchedule()
+	options := wasmer.CompilationOptions{
+		GasLimit:           gasLimit,
+		UnmeteredLocals:    uint64(gasSchedule.WASMOpcodeCost.LocalsUnmetered),
+		OpcodeTrace:        false,
+		Metering:           true,
+		RuntimeBreakpoints: true,
+	}
+	newInstance, err := wasmer.NewInstanceFromCompiledCodeWithOptions(compiledCode, options)
+	if err != nil {
+		log.Trace("NewInstanceFromCompiledCodeWithOptions", "error", err)
+		return false
+	}
+
+	context.instance = newInstance
+
+	idContext := arwen.AddHostContext(context.host)
+	context.instance.SetContextData(idContext)
+
+	return true
+}
+
+func (context *runtimeContext) StartWasmerInstance(contract []byte, gasLimit uint64, newCode bool) error {
+	if context.RunningInstancesCount() >= context.maxWasmerInstances {
+		context.instance = nil
+		return arwen.ErrMaxInstancesReached
+	}
+
+	blockchain := context.host.Blockchain()
+	codeHash := blockchain.GetCodeHash(context.GetSCAddress())
+
+	warmInstanceSet := context.setWarmInstanceWhenNeeded()
+	if warmInstanceSet {
+		return nil
+	}
+
+	compiledCodeSet := context.makeInstanceFromCompiledCode(codeHash, gasLimit, newCode)
+	if compiledCodeSet {
+		return nil
+	}
+
+	log.Trace("Creating a new Wasmer instance")
```

Comment:
Optional: the rest of this function, starting here, could be put into a separate private method `makeInstanceFromContractBytecode()`.

Reply:
done